### PR TITLE
Exploratory PBMCs

### PR DESCRIPTION
**Pull request by MikeLippincott**
Busy PR, a lot going on at once. This is PR 1 of 2 to set up this repo for PBMC analysis

### CODE REVIEW

File: 1.Exploratory_Data_Analysis/exploratory_analysis.sh
Comment (roshankern) [inline 1.Exploratory_Data_Analysis/exploratory_analysis.sh:1]: Same comment as other `sbatch` file, consider adding documentation to explain what this code does.

File: 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+get_ipython().run_line_magic("matplotlib", "inline")
+import itertools
+import pathlib
+
+import pyarrow.parquet as pq
+from sklearn.cluster import KMeans
+from sklearn.manifold import TSNE
+
+sys.path.append("..")
+from utils.utils import df_stats, plot_features_all_cluster
```
Comment (roshankern) [inline 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py:23]: A small thing here but the [PEP style guide](https://peps.python.org/pep-0008/#imports) says:

> Imports should be grouped in the following order:
> Standard library imports.
> Related third party imports.
> Local application/library specific imports.

So consider moving the `itertools` and `pathlib` imports up unless I am missing something here.

File: 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+
+celltype = "SHSY5Y"
+
+
+# In[3]:
+
+
+# Parameters
+celltype = "PBMC"
```
Comment (roshankern) [inline 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py:35]: Did you mean to define `celltype` twice here? This seems like unnecessary code.

File: 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py
Comment (roshankern) [inline 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py:1]: Same comments as `0.cell_count_analysis.py` (imports and defining `celltype` twice).

File: 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    x="umap_1",
+    y="umap_2",
+    hue="Treatment",
+    legend="full",
+    alpha=0.7,
+)
+plt.title("Visualized on umap")
+plt.legend(bbox_to_anchor=(1.02, 1), loc="upper left", borderaxespad=0)
+
+# if path does not exist create it
```
Comment (roshankern) [inline 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py:117]: The scope of the 2 comments in the highlighted code is a bit confusing as they each have an empty line underneath. Consider moving the comments or changing the whitespace to clarify what the comments refer to.

File: 1.Exploratory_Data_Analysis/scripts/umap_analysis_plate2.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+
+# In[3]:
+
+
+# Parameters
+celltype = "SHSY5Y"
+
+
+# <span id="papermill-error-cell" style="color:red; font-family:Helvetica Neue, Helvetica, Arial, sans-serif; font-size:2em;">Execution using papermill encountered an exception here and stopped:</span>
```
Comment (roshankern) [inline 1.Exploratory_Data_Analysis/scripts/umap_analysis_plate2.py:42]: Should this comment be included here? Or maybe the notebook needs to be rerun?

File: 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import pyarrow as pa
+import pyarrow.parquet as pq
+import seaborn as sns
+from matplotlib import pyplot as plt
+from pycytominer.cyto_utils import infer_cp_features
+from sklearn.linear_model import LinearRegression
+from sklearn.preprocessing import LabelEncoder
+
+sys.path.append("..")
+# from ..utils.utils import df_stats
```
Comment (roshankern) [inline 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:22]: Can this line be removed?

File: 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+# Here I run the same analysis as above but with dosage of a treatment being a factor in the linear model. All features and treatments will be exported into 1 file.
+#
+# Linear Model:
+# $y = \beta _{0}x+ \beta _{1}x+ \beta _{2}x+ \epsilon$ where;
+# $y$ is each feature
+# $x$ is the inputed variables
+# $\beta _{0}$ is the beta coefficient attributed to cell count.
+# $\beta _{1}$ is the beta coefficient attributed to Inducer, Inhibitor, and Inhibitor Dose.
+# $\beta _{2}$ is the beta coefficient attributed to Inducer dose.
+# $\epsilon$ is the residual variance not explained by factors in the model
```
Comment (roshankern) [inline 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:78]: Consider changing this block to use multi-line comment string: """comment"""

File: 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+# In[6]:
+
+
+# Loop for each treatment then each feature
+
+# define the control and treatment
+# Setup linear modeling framework
```
Comment (roshankern) [inline 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:86]: It make make more sense to reorder these comments. Ex: move `Loop for each treatment then each feature` to right above the `for treatment in features_df...` code.

File: 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            "Treatment_Dose",
+        ),
+    ]
+
+    tmp_df["twob_Metadata_Treatment_Inhibitor_Dose"] = LabelEncoder().fit_transform(
+        tmp_df["twob_Metadata_Treatment_Inhibitor_Dose"]
+    )
+    tmp_df["Treatment_Dose"] = LabelEncoder().fit_transform(tmp_df["Treatment_Dose"])
+
+    X = pd.concat([df.loc[:, model_covariates], tmp_df], axis=1)
```
Comment (roshankern) [inline 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:113]: Consider using a more descriptive variable than `X`

Comment (roshankern) [reply 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:113]: After looking at the rest of the code it seems like you just wanted to match the parameter name so I think this is an okay name, but consider adding a comment above this line to explain what is happening/what X represents.

File: env.yaml [hunk 9]
```diff
@@ -13,6 +13,8 @@ dependencies:
   - pytorch
   - seaborn
   - pyarrow
+  - numba
```
Comment (roshankern) [inline env.yaml:16]: Consider specifying which channels these packages should come from as well.

**Review by roshankern — APPROVED** [review]
A couple small comments you can address or ignore, then LGTM!

For large PR's similar to this that you need to complete in the future, please consider following the `Large Refactors` strategy described in [Waylab Code Strategy](https://docs.google.com/document/d/1htu3DQPZebH2mv-Wyxa45MzUz6Mt8iXZ96t_O0hSSdM/edit)

File: 0.Preprocess_Data/pre-processing.sh
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 0.Preprocess_Data/pre-processing.sh:1]: This is fairly new to the Way Lab. I agree, noting what these `sbatch` commands do would be nice for readability.
We currently have a private document for Way Lab playbooks for `slurm` clusters but we should link a public repo `README` that explains what these do

File: 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+# subset data frame to 1000 samples too much data results in poor clustering
+df = df1.sample(n=1000)
+# save memory by deleting df1
+del df1
+
+
+# In[5]:
+
+
+# Code snippet for metadata extraction by Jenna Tomkinson
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py:56]: Credit given where credit is due!

File: 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py (hunk 6 above)
Comment (MikeLippincott) [reply 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:78]: This is a markdown cell in the notebook. Is there a way tot specify in the converted `.py` file?

File: 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py (hunk 1 above)
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py:23]: fixed

File: 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py (hunk 2 above)
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/0.cell_count_analysis.py:35]: I did not mean to, but I am removing it now

File: 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py (hunk 3 above)
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/1.umap_analysis_plate2.py:117]: Applied Changes

File: 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py (hunk 5 above)
Comment (MikeLippincott) [reply 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:22]: yes, removing

File: env.yaml (hunk 9 above)
Comment (MikeLippincott) [reply env.yaml:16]: resolved

File: 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py (hunk 6 above)
Comment (roshankern) [reply 3.sc_Morphology_Linear_Modeling/scripts/1b.fit_linear_model_2beta.py:78]: Just saw this. I don't think there is a way to specify but if this how jupyter convert output looks then I think it is fine.

**Pull request merged**